opencog/ure/forwardchainer/ForwardChainer.cc

```diff
@@ -411,6 +412,28 @@ HandleSet ForwardChainer::apply_rule(const Rule& rule)
 		AtomSpace derived_rule_as(&ref_as);
 		Handle rhcpy = derived_rule_as.add_atom(rule.get_rule());
 
+		HandleSeq clauses = rule.get_clauses();
+		Handle vdecl = rule.get_vardecl();
+		HandleSet varset1;
+		HandleSet varset2;
+
+		if (vdecl != Handle::UNDEFINED)
+		{
+			if (VariableListCast(vdecl))
+				varset1 = VariableListCast(vdecl)->get_variables().varset;
+			else
+				varset1.insert(vdecl);
+
+			for (auto var : varset1)
+				if (var->get_type() == TYPED_VARIABLE_LINK)
+					varset2.insert(var->getOutgoingAtom(0));
+		}
+
+		for (Handle clause : clauses)
+			if (is_constant(varset2,clause))
```

Comment:
If `vdecl` is `Handle::UNDEFINED`, I don't think this works.

Note that `vdecl == Handle::UNDEFINED` is different than the empty variable declaration `(VariableList)`. The latter means that there are no variable, while the former means that the variable declaration is the default one, and will be automatically inferred by the scope link.